internal/indexer/postgres/indexmanifest.go

```diff
@@ -0,0 +1,158 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/rs/zerolog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/pkg/microbatch"
+)
+
+func indexManifest(ctx context.Context, pool *pgxpool.Pool, ir *claircore.IndexReport) error {
+	const (
+		query = `
+		INSERT INTO manifest_index(package_id, dist_id, repo_id, manifest_hash)
+		VALUES ($1, $2, $3, $4)
+		ON CONFLICT DO NOTHING;
+		`
+	)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "internal/indexer/postgres/indexManifest").
+		Logger()
+
+	if ir.Hash.String() == "" {
+		return fmt.Errorf("received empty hash. cannot associate contents with a manifest hash")
+	}
+	hash := ir.Hash.String()
+
+	records := ir.IndexRecords()
+	if len(records) == 0 {
+		log.Warn().Msg("manifest being indexed has 0 scan records")
+		return nil
+	}
+
+	// obtain a transaction scopped batch
+	tx, err := pool.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("store:indexDistributions failed to create transaction: %v", err)
+	}
+	defer tx.Rollback(ctx)
+
+	queryStmt, err := tx.Prepare(ctx, "queryStmt", query)
+	if err != nil {
+		return fmt.Errorf("failed to create statement: %v", err)
+	}
+
+	mBatcher := microbatch.NewInsert(tx, 500, time.Minute)
+	var ignored int
+	for _, record := range records {
+		// ignore nil packages
+		if record.Package == nil {
+			ignored++
+			continue
+		}
+
+		v, err := toValues(*record)
+		if err != nil {
+			return fmt.Errorf("received a record with an invalid id: %v", err)
+		}
+
+		// if source package exists create record
+		if v[0] != nil {
+			err = mBatcher.Queue(
+				ctx,
+				queryStmt.SQL,
+				v[0],
+				v[2],
+				v[3],
+				hash,
+			)
+			if err != nil {
+				return fmt.Errorf("batch insert failed for source package record %v: %v", record, err)
+			}
+		}
+
+		err = mBatcher.Queue(
+			ctx,
+			queryStmt.SQL,
+			v[1],
+			v[2],
+			v[3],
+			hash,
+		)
+		if err != nil {
+			return fmt.Errorf("batch insert failed for package record %v: %v", record, err)
+		}
+
+	}
+	err = mBatcher.Done(ctx)
+	if err != nil {
+		return fmt.Errorf("final batch insert failed: %v", err)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit tx: %v", err)
+	}
+	return nil
+}
+
+// toValues is a helper method which checks for
+// nil pointers inside an IndexRecord before
+// returning an associated pointer to the artifact
+// in question.
+//
+// v[0] source package id or nil
+// v[1] package id or nil
+// v[2] distribution id or nil
+// v[3] repository id or nil
+func toValues(r claircore.IndexRecord) ([4]*uint64, error) {
```

Comment:
This should take the pointer to the record, instead of making a copy.

Reply:
@hdonnay since the method that calls this creates the IR (it is not passed in from an even higher method) I was trying to avoid passing pointers and completely use the stack as a general "don't gc as much".

Reply:
but you're having to dereference a pointer to call this, which came out of a returned slice of pointers. the value is almost certainly on the heap

Reply:
crap, got this confused with protoRecord method, nvm the record is already a pointer when it enters indexManifest.

Reply:
I'm going to leave this copy, as its being used in the affectedManifest method with the attention described above.